functional/FunctionalTest.java

```diff
@@ -3761,6 +3764,309 @@ public static void getDefaultRetention_test() throws Exception {
+                .object(objectName)
+                .config(emptyConfig)
+                .bypassGovernanceMode(true)
+                .build());
+
+      } finally {
+        client.removeObject(
+            RemoveObjectArgs.builder().bucket(bucketName).object(objectName).build());
+        client.removeBucket(RemoveBucketArgs.builder().bucket(bucketName).build());
+      }
+
+      mintSuccessLog("setObjectRetention(SetObjectRetentionArgs args)", null, startTime);
+
+    } catch (Exception e) {
+      ErrorResponse errorResponse = null;
+      if (e instanceof ErrorResponseException) {
+        ErrorResponseException exp = (ErrorResponseException) e;
+        errorResponse = exp.errorResponse();
+      }
+
+      // Ignore NotImplemented error
+      if (errorResponse != null && errorResponse.errorCode() == ErrorCode.NOT_IMPLEMENTED) {
+        mintIgnoredLog("setObjectRetention(SetObjectRetentionArgs args)", null, startTime);
+      } else {
+        mintFailedLog(
+            "setObjectRetention(SetObjectRetentionArgs args)",
+            null,
+            startTime,
+            null,
+            e.toString() + " >>> " + Arrays.toString(e.getStackTrace()));
+        throw e;
+      }
+    }
+  }
+
+  /** Test: getObjectRetention(GetObjectRetentionArgs args). */
+  public static void getObjectRetention_test1() throws Exception {
+    if (!mintEnv) {
+      System.out.println("Test: getObjectRetention(GetObjectRetentionArgs args)");
+    }
+
+    long startTime = System.currentTimeMillis();
+    String bucketName = getRandomName();
+    String objectName = getRandomName();
+    try {
+      client.makeBucket(MakeBucketArgs.builder().bucket(bucketName).objectLock(true).build());
+      try {
+        try (final InputStream is = new ContentInputStream(1 * KB)) {
+          PutObjectOptions options = new PutObjectOptions(1 * KB, -1);
+          options.setContentType(customContentType);
+          client.putObject(bucketName, objectName, is, options);
+        }
+
+        ZonedDateTime retentionUntil = ZonedDateTime.now().plusDays(3);
+        Retention expectedConfig = new Retention(RetentionMode.GOVERNANCE, retentionUntil);
+        client.setObjectRetention(
+            SetObjectRetentionArgs.builder()
+                .bucket(bucketName)
+                .object(objectName)
+                .config(expectedConfig)
+                .build());
+
+        Retention config =
+            client.getObjectRetention(
+                GetObjectRetentionArgs.builder().bucket(bucketName).object(objectName).build());
+
+        if ((!(config
+                .retainUntilDate()
+                .toString()
+                .substring(0, config.retainUntilDate().toString().indexOf("T"))
+                .equals(
+                    expectedConfig
+                        .retainUntilDate()
+                        .toString()
+                        .substring(0, expectedConfig.retainUntilDate().toString().indexOf("T")))))
```

Comment:
1. Too many nesting.
2. Isn't `ZonedDateTime.equals()` working?

Reply:
`ZonedDateTime.equa` didn't work as the value returned by server is `2020-05-30T08:51:38Z` while the value set is `2020-05-30T14:21:38.951+05:30[Asia/Kolkata]`.

Reply:
Thanks for you suggestion. 
`ZonedDateTime retentionUntil = ZonedDateTime.now(Time.UTC).plusDays(3).withNano(0);` works.